blockchain/blockchain.go

```diff
@@ -696,6 +705,90 @@ func (bc *blockchain) MintNewBlock(
 	return blk, nil
 }
 
+func (bc *blockchain) MintNewBlockWithActionIterator(
+	actionMap map[string][]action.SealedEnvelope,
+	producer *iotxaddress.Address,
+	dkgAddress *iotxaddress.DKGAddress,
+	seed []byte,
+	data string,
+) (*Block, error) {
+	bc.mu.RLock()
+	defer bc.mu.RUnlock()
+
+	// Use block height as the nonce for coinbase transfer
+	cb := action.NewCoinBaseTransfer(bc.tipHeight+1, bc.genesis.BlockReward, producer.RawAddress)
+	bd := action.EnvelopeBuilder{}
+	// TODO the nonce is wrong, if bd also submit actions
+	elp := bd.SetNonce(bc.tipHeight + 1).
+		SetDestinationAddress(producer.RawAddress).
+		SetGasLimit(protocol.GasLimit).
+		SetAction(cb).Build()
+	selp, err := action.Sign(elp, producer.RawAddress, producer.PrivateKey)
+
+	// run execution and update state trie root hash
+	ws, err := bc.sf.NewWorkingSet()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to obtain working set from state factory")
+	}
+	blkHeight := bc.tipHeight + 1
+	blkHash := bc.tipHash
+	producerPubKey := producer.PublicKey
+	blkTimestamp := bc.now()
+	pkHash := keypair.HashPubKey(producerPubKey)
+	addr := address.New(bc.config.Chain.ID, pkHash[:])
+	producerAddr := addr.IotxAddress()
+	gasLimit := GasLimit
+
+	// update state factory
+	ctx := protocol.WithRunActionsCtx(context.Background(),
+		protocol.RunActionsCtx{
+			BlockHeight:     blkHeight,
+			BlockHash:       blkHash,
+			ProducerPubKey:  producerPubKey,
+			BlockTimeStamp:  time.Unix(int64(blkTimestamp), 0).Unix(),
+			ProducerAddr:    producerAddr,
+			GasLimit:        &gasLimit,
+			EnableGasCharge: bc.config.Chain.EnableGasCharge,
+		})
+	actionIterator := actioniterator.NewActionIterator(actionMap)
+	appliedActionList, receiptMap, err := PickAction(ctx, ws, bc, &gasLimit, actionIterator)
+	// include coinbase transfer
+	appliedActionList = append(appliedActionList, selp)
+
+	// initial a block with empty actions
+	blk := NewBlock(bc.config.Chain.ID, blkHeight, blkHash, blkTimestamp,
+		producerPubKey, appliedActionList)
+	blk.Header.DKGID = []byte{}
+	blk.Header.DKGPubkey = []byte{}
+	blk.Header.DKGBlockSig = []byte{}
+	if dkgAddress != nil && len(dkgAddress.PublicKey) > 0 && len(dkgAddress.PrivateKey) > 0 && len(dkgAddress.ID) > 0 {
+		blk.Header.DKGID = dkgAddress.ID
+		blk.Header.DKGPubkey = dkgAddress.PublicKey
+		var err error
+		if _, blk.Header.DKGBlockSig, err = crypto.BLS.SignShare(dkgAddress.PrivateKey, seed); err != nil {
+			return nil, errors.Wrap(err, "Failed to do DKG sign")
+		}
+	}
+	blk.Receipts = make(map[hash.Hash32B]*action.Receipt)
+	if bc.sf == nil {
+		return nil, errors.New("statefactory cannot be nil")
+	}
+	for k, v := range receiptMap {
+		blk.Receipts[k] = v
+	}
+	root, err := bc.runActionsWithActionIterator(ctx, blk, ws, false)
```

Comment:
runActionsWithActionIterator will execute contract again. It seems to be conflicting